### PR TITLE
Updating for IAM Proxy

### PR DESCRIPTION
**Pull request by matthewdfuller**
- The `init` script runs the `iam-proxy.sh` util script first
- The `iam-proxy.sh` util script sets up the IPTABLES proxy routing and submits the `iam-proxy.service` service unit
- The `iam-proxy.service` service unit pulls the proxy docker container and runs it.
- Additional updates are made to the `setup-credentials.sh` script to pass the container role (created in `infrastructure` repo to the secrets downloader container.

### CODE REVIEW

**Comment by matthewdfuller** [comment]
Not yet tested with infrastructure; DO NOT MERGE

File: init [hunk 1]
```diff
@@ -20,6 +20,11 @@ if [ "$NODE_ROLE" = "it-hybrid" ]; then
   sudo ${SCRIPTDIR}/itv1/util/etcd2-setup.sh $SCRIPTDIR
```
Comment (chr0n1x) [inline init:20]: the `etcd2` setup scripts run a docker container

Comment (chr0n1x) [inline init:20]: not sure if that affects any of this?

File: v3/util/iam-proxy.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+source /etc/environment
+
+if [ "${NODE_ROLE}" != "worker" ]; then
+    exit 0
+fi
+
+etcdctl set /images/iam-proxy  	 "behance/iam-docker:latest"
+docker pull $(etcdctl get /images/iam-proxy)
```
Comment (chr0n1x) [inline v3/util/iam-proxy.sh]: lol this seems kind of silly

File: init (hunk 1 above)
Comment (matthewdfuller) [inline init:20]: This only has to run on worker. These etcd setups only run on control and it.

File: v3/util/iam-proxy.sh (hunk 2 above)
Comment (matthewdfuller) [inline v3/util/iam-proxy.sh]: Yeah - just copying the existing format from images.sh though

File: v3/util/iam-proxy.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+export INTERFACE="docker0"
+
+# These will not work until Docker > 1.9 is running
+#export GATEWAY="$(docker network inspect "$NETWORK" | grep Gateway | cut -d '"' -f 4)"
+#export INTERFACE="br-$(docker network inspect "$NETWORK" | grep Id | cut -d '"' -f 4 | head -c 12)"
+
+sudo iptables -t nat -I PREROUTING -p tcp -d 169.254.169.254 --dport 80 -j DNAT --to-destination "$GATEWAY":8080 -i "$INTERFACE"
+
+# Sometimes this is needed?
+systemctl daemon-reload
```
Comment (chr0n1x) [inline v3/util/iam-proxy.sh]: shouldn't this be happening after you copy it into `etc`?

File: v3/util/iam-proxy.sh (hunk 2 above)
Comment (chr0n1x) [inline v3/util/iam-proxy.sh]: I see why it's needed, just kind of topsy-turkey

File: v3/util-units/iam-proxy.service [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  --net=host \
+  $($IMAGE)
+
+ExecStop=-/usr/bin/docker stop iam-proxy
+
+[Install]
+WantedBy=multi-user.target
+
+[X-Fleet]
+Global=false
```
Comment (chr0n1x) [inline v3/util-units/iam-proxy.service]: we only need this unit to run on a _single_ worker node?

File: init [hunk 5]
```diff
@@ -20,6 +20,11 @@ if [ "$NODE_ROLE" = "it-hybrid" ]; then
...
 fi
 
+# Worker tier - must run the IAM proxy setup before any other containers
+if [ "$NODE_ROLE" = "worker" ] && [ "$VERSION" = "v3" ]; then
```
Comment (phylake) [inline init]: Forgetting to update this when we hit v4 seems like an easy thing to do. Is there an alternative?

Comment (chr0n1x) [inline init]: checking for the existence of the file works too

File: v3/util/iam-proxy.sh (hunk 3 above)
Comment (matthewdfuller) [inline v3/util/iam-proxy.sh]: @chr0n1x maybe? https://github.com/adobe-platform/mesos-systemd/blob/master/v3/util/etcd2-setup.sh#L20
We should change in both places if so, this was just a copy pasta

File: v3/util-units/iam-proxy.service (hunk 4 above)
Comment (matthewdfuller) [inline v3/util-units/iam-proxy.service]: Good catch, fixing

File: v3/util/iam-proxy.sh (hunk 3 above)
Comment (chr0n1x) [inline v3/util/iam-proxy.sh]: oh lol weird. @nBerg did this - what happened when it _was_ needed?

File: v3/util/iam-proxy.sh [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+source /etc/environment
+
+if [ "${NODE_ROLE}" != "worker" ]; then
+    exit 0
+fi
+
+docker pull "behance/iam-docker:latest"
```
Comment (bryanlatten) [inline v3/util/iam-proxy.sh:11]: we should really start adding the full-qualified registry in these "naked pulls" as we start down the ECR rabbit hole (or make it a parameter)

**Comment by cdmurph32** [comment]
How would an app specify it's IAM role?  Does there need to be a change to the ImageDefinition table and FD API?

**Comment by matthewdfuller** [comment]
@cdmurph32 yes FD needs to be updated to support running docker containers with "docker run {{role label here}}..". Although this does not break anything existing since apps launched on workers do not have any additional permissions (beyond what the worker instance had)

File: v3/opt/ecr/aws-ecr-login.service [hunk 7]
```diff
@@ -7,6 +7,7 @@ After=network-online.target
 
 [Service]
 Type=oneshot
+EnvironmentFile=/etc/environment
```
Comment (cdmurph32) [inline v3/opt/ecr/aws-ecr-login.service:10]: Where are the contents of /etc/environment set?

Comment (matthewdfuller) [inline v3/opt/ecr/aws-ecr-login.service:10]: That file was written as part of the cloud-config:
https://github.com/adobe-platform/infrastructure/tree/master/v2/src/userdata/templates/write_files/environment
Specifically:
https://github.com/adobe-platform/infrastructure/blob/master/v2/src/userdata/templates/write_files/environment/worker.yml#L2

**Comment by chr0n1x** [comment]
👍

File: v3/util/iam-proxy.sh [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+sudo iptables -t nat -I PREROUTING -p tcp -d 169.254.169.254 --dport 80 -j DNAT --to-destination "$GATEWAY":8080 -i "$INTERFACE"
+
+sudo fleetctl submit "${SCRIPTDIR}/../util-units/iam-proxy.service"
+sudo fleetctl start "iam-proxy.service"
+
+# sudo cp "${SCRIPTDIR}/../util-units/iam-proxy.service" /etc/systemd/system/
+# sudo systemctl start iam-proxy.service
+
+# Wait until service is active
+until [ "`/usr/bin/docker inspect -f {{.State.Running}} iam-proxy`" == "true" ]; do
```
Comment (ko-be) [inline v3/util/iam-proxy.sh:30]: This loop can theoretically go into an infinite loop.  @matthewdfuller

File: v3/util-units/iam-proxy.service [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  behance/iam-docker:latest"
+
+ExecStop=-/usr/bin/docker stop iam-proxy
+
+[Install]
+WantedBy=multi-user.target
+
+[X-Fleet]
+Global=true
+MachineMetadata=role=worker
```
Comment (ko-be) [inline v3/util-units/iam-proxy.service:32]: nitpick, missing newline at end of file

File: v3/util/iam-proxy.sh [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+	echo "Waiting for iam-proxy service..."
+    sleep 5;
+done;
+
+exit 0
```
Comment (ko-be) [inline v3/util/iam-proxy.sh]: nitpick, missing newline at end of file

File: v3/util/iam-proxy.sh (hunk 8 above)
Comment (matthewdfuller) [inline v3/util/iam-proxy.sh:30]: @ko-be yep - if IAM-proxy doesn't start up, the machine is hosed anyway.

File: v3/util/iam-proxy.sh (hunk 6 above)
Comment (matthewdfuller) [inline v3/util/iam-proxy.sh:11]: @bryanlatten agreed, I will make a card to update them all throughout the scripts together

**Comment by bossjones** [comment]
per IRL discussion about this PR, LGTM 👍

**Comment by ko-be** [comment]
Aside from the missing newlines, 👍

File: v3/util-units/iam-proxy.service (hunk 9 above)
Comment (chr0n1x) [inline v3/util-units/iam-proxy.service:32]: there shouldn't be one

**Pull request merged**